src/upload-button.tsx

```diff
@@ -218,6 +253,37 @@ export const UploadButton = ({
                 };
             });
 
+            if (owner !== null) {
+                // The cockpit.file() API does not support setting an owner/group when uploading a new
+                // file with fsreplace1. This requires a re-design of the Files API:
+                // https://issues.redhat.com/browse/COCKPIT-1215
+                //
+                // For now we create an empty file using fsreplace1 and give it the proper ownership and using that tag
+                // upload the to be uploaded file. This prevents uploading with the wrong ownership.
+                //
+                // To support changing permissions after upload with our change permissions dialog we obtain the
+                // file mode using `stat` as fsreplace1 does not report this back except via the `tag` which is not
+                // a stable interface.
+                try {
+                    await cockpit.file(destination, { superuser: "try" }).replace(""); // TODO: cleanup might not work?
+                    await cockpit.spawn(["chown", owner, destination], { superuser: "try" });
+                    await cockpit.file(destination, { superuser: "try" }).read()
+                            .then((((_content: string, tag: string) => {
+                                options = { superuser: "try", tag };
+                            }) as any /* eslint-disable-line @typescript-eslint/no-explicit-any */));
+                    const stat = await cockpit.spawn(["stat", "--format", "%a", destination], { superuser: "try" });
+                    fileModes.push(Number.parseInt(stat.trimEnd(), 8));
+                } catch (exc) {
+                    await cockpit.spawn(["rm", "-f", destination], { superuser: "require" });
+                    console.warn("Cannot set initial file permissions", exc);
+                    addAlert(_("Cancelled"), AlertVariant.warning, "upload",
+                             cockpit.format(_("Cancelled upload of $0"), file.name));
+                    cancelledUploads.push(file);
+                    return;
```

Comment:
These 7 added lines are not executed by any test.

Reply:
Not entirely sure how to test this, I'll give it some more thought.